### DISCUSSION TITLE
Iterator comparison question

### DISCUSSION THREAD

**Discussion by andrewwongagility** [foxglove/mcap#1123] [General] [answered]
I noticed that the code: `LinearMessageView::Iterator a == messageView.begin()` always seems to returns false. For example, when reading an mcap file in the below code snippet:

```cpp
  auto messageView = reader.readMessages();
  for (auto it = messageView.begin(); it != messageView.end(); it++) {
    if (it == messageView.begin()){
      std::cout << "AT START!\n";
    }
    std::cout << "it == messageView.begin()\t" << std::boolalpha << (it == messageView.begin()) << std::endl;
    ...
   }
```
The output is:
```txt
INFO: Build completed successfully, 3 total actions
it == messageView.begin()       false
it == messageView.begin()       false
it == messageView.begin()       false
...
```
I'm wondering if this is expected behavior? I had expected "AT START" to be printed, and the first row to print `true`. 

Looking [here](https://github.com/foxglove/mcap/blob/main/cpp/mcap/include/mcap/reader.inl#L1754), I noticed that the operator overload for `==` checks whether two `unique_ptrs` are equal, instead of the objects the pointers are pointing to. I wanted to check if this was intended behavior.

I'm not sure if I'm missing something. Thank you very much for any help!

**Comment by jtbandes** [comment]
Agreed that this seems like a bug/non-ideal behavior.

Internal tracking issue: https://linear.app/foxglove/issue/FG-4914/linearmessageview-begin-does-not-compare-equal-to-begin

**Comment by andrewwongagility** [comment, marked as answer]
I think this has been fixed in the PR: https://github.com/foxglove/mcap/pull/969
Thank you for the quick fix!
